src/Workspaces/Core/Portable/Workspace/ProjectSystem/FileWatchedPortableExecutableReferenceFactory.cs

```diff
@@ -31,7 +29,7 @@ internal sealed class FileWatchedPortableExecutableReferenceFactory
     /// File watching tokens from <see cref="_fileReferenceChangeContext"/> that are watching metadata references. These are only created once we are actually applying a batch because
     /// we don't determine until the batch is applied if the file reference will actually be a file reference or it'll be a converted project reference.
     /// </summary>
-    private readonly Dictionary<PortableExecutableReference, IWatchedFile> _metadataReferenceFileWatchingTokens = [];
+    private readonly Dictionary<PortableExecutableReference, (IWatchedFile Token, int RefCount)> _metadataReferenceFileWatchingTokens = [];
```

Comment:
💡 This is what ReferenceCountedDisposable is for.

Reply:
Oops missed this before I hit the button.  I tried using `ReferenceCountedDisposable` but I couldn't seem to make the usage very ergonomic.  No one is really taking the watcher token inside a `using` statement, its just added to the map.

I couldn't figure out a simple pattern to just increment and decrement the counter so went with this.

Reply:
I'm ok with the merge as-is, as this is nicely self contained.  @sharwell We'd def love to see how to use RCD effectively here.  If you want to make a pr here, that would be great.  Thanks!